### ISSUE TITLE
Warning: Avoid app logic that relies on enumerating keys on a component instance. The keys will be empty in production mode to avoid performance overhead.

### ISSUE THREAD

**Issue by frasza** [nuxt-modules/i18n#2856]
### Environment

------------------------------
- Operating System: Windows_NT
- Node Version:     v21.6.2
- Nuxt Version:     3.11.0
- CLI Version:      3.10.1
- Nitro Version:    2.9.4
- Package Manager:  npm@10.2.4
- Builder:          -
- User Config:      modules, devtools
- Runtime Modules:  @nuxtjs/i18n@8.2.0
- Build Modules:    -
------------------------------

### Reproduction

https://github.com/frasza/i18n-warn

Spin up dev server, open dev tools and click on NuxtLink. Warning appears.

### Describe the bug

Recently I noticed warning popup in my Nuxt project and I have tried to set up fresh Nuxt project with i18n dependacy and I still get warning when for navigating across pages.

Warning:
```
[Vue warn]: Avoid app logic that relies on enumerating keys on a component instance. The keys will be empty in production mode to avoid performance overhead.
```

Noticed that I get warning the moment I add i18n to the modules.

**Comment by BobbieGoede** [comment]
I'm unable to reproduce this using your reproduction, I did have to update Nuxt to 3.11.1 or the dev server wouldn't run. Has this already been resolved?

**Comment by frasza** [comment]
I still have problem with warning even tho I am on Nuxt 3.11.1.

**Comment by BobbieGoede** [comment]
Do you get the warning using this updated reproduction? https://stackblitz.com/edit/frasza-i18n-warn-76cwcr?file=nuxt.config.ts

**Comment by frasza** [comment]
Yeah, I have downloaded the project, run it locally and I still get it 🤔
![image](https://github.com/nuxt-modules/i18n/assets/22049682/f91b4183-6d64-49a4-b1b4-e289e41d7b41)

**Comment by BobbieGoede** [comment]
Ah after running it locally I get the same warning, looks like this happens when you add Nuxt I18n without configuring any locales. The module tries to navigate/redirect to a localized page but there are none, maybe we should add some checks to disable the module if nothing is configured.

**Comment by frasza** [comment]
Locally I have it configured tho.

```ts
i18n: {
    vueI18n: '@/i18n.config.ts',
  },
```

```ts
import en from '@/locales/en.json'
import si from '@/locales/si.json'

export default defineI18nConfig(() => ({
  legacy: false,
  locale: 'si',
  messages: {
    si,
    en,
  },
}))
```

Perhaps something here wrong as well...

**Comment by BobbieGoede** [comment]
Let me know if these changes fix it for you:

```diff
i18n: {
    vueI18n: '@/i18n.config.ts',
+   locales: ['en', 'si']
+   defaultLocale: ['si']
},
```

```diff
import en from '@/locales/en.json'
import si from '@/locales/si.json'

export default defineI18nConfig(() => ({
   legacy: false,
-  locale: 'si',
   messages: {
     si,
     en,
   },
}))
```

**Comment by frasza** [comment]
I see this forces locale in URL paht (e.g.: /si). How can I escape this, since we are not using locales in url pahts?

**Comment by BobbieGoede** [comment]
Check out the docs on routing strategies https://i18n.nuxtjs.org/docs/guide

**Comment by frasza** [comment]
Thank you.

Indeed! In this case I do not get warning. Now it is a bit 'weird' that configuration is split between the files but yeah, it seems to be working well.

**Comment by adgower** [comment]
I get "no match found for location with path "/es"

```
i18n: {
    vueI18n: "./il8n.config.ts",
    strategy: "no_prefix",
    locales: ["en", "es"],
  },
```

**Comment by itimers** [comment]
When i use "@nuxtjs/i18n": "^8.3.1", there is no warn but if update on newer vesrion i got same warning

**Comment by wug-ge** [comment]
I was struggling with this bug and can confirm it is only happening in v8, upgrading to v9 resolves this issue.

What causes this warning:
- The router is trying to lookup routes with a certain suffix, e.g. "home__en" even when nothing like that is configured and only a route named "home" exists.
- This causes a route error and the following error getting thrown: https://github.com/vuejs/router/blob/main/packages/router/src/errors.ts#L86
- nuxt-i18n seems to have a quiet complex route object which includes a property called "instances" having a ProxyObject, running JSON.stringify during the error causes Vue.js runtime core's .ownKeys trap to be called which outputs the following warning.

Let me know if you need any more details, most important is probably though that it is resolved in newer versions. I suspect some issue with the prefix strategy configuration.

**Comment by BobbieGoede** [comment]
@wug-ge 
Thank you for investigating this issue! 

It's good to hear that this seems to be resolved in v9, I have done a few refactors that made it possible to improve the correctness of route resolution in nuxt-i18n. Unfortunately this makes it difficult to pinpoint what change actually resolved the issue to backport it.

You're right that it's likely related to the prefix strategy, it could also be caused by trying to resolve routes using paths instead of route names (e.g. `localePath('/about-us')` instead of either `localePath('about-us')` or `localePath({ name: 'about-us' })`).

**Comment by wug-ge** [comment]
@BobbieGoede In my testing, using a path instead of a route name seemed to solve this problem actually, however I can't confirm if this is always the case. I can test later if the issue is related to nuxt-i18n or vue-router/vue-i18n itself / what exact version is resolving this issue. 

The reproduction @frasza build is very useful for testing.

**Comment by BobbieGoede** [comment]
@wug-ge 
If it seems like paths are working better you're likely using a catch-all route in your project, in which case the route resolution resolves to a path but internally it matched the catch-all route (not transparent with `localePath` as it returns a string).

Using route names is the easiest way for nuxt-i18n to find the localized equivalent of a route.

**Comment by wug-ge** [comment]
@BobbieGoede 
I've tested a bit more, and can confirm the issue exists in v8.5.6 and is removed in v9.0.0.
The proper solution for it is probably using the 'no_prefix' strategy, this is what I actually needed and maybe @frasza as well.

The issue is caused by misconfiguration, if we take a look at the reproduction: no strategy is configured, causing the strategy to be 'prefix_except_default'.
However also no defaultLocale is configured which means the route is expected to have a prefix.

The NuxtLink is also not using useLocalePath(), which means  a route is trying to be accessed that does not exist. The only issue is maybe that it still somehow works instead of failing given no defaultLocale is configured. (or having a proper warning for that case since the warning we get is very misleading)

Or that (contrary to what I though) the default strategy is 'prefix_except_default' while 'no_prefix' might be better?

Just a few ideas, but given it is certainly caused by a misconfiguration, I'm fine with closing this issue as well. Having a more reasonable warning would be nice though.

**Comment by BobbieGoede** [comment]
I have added a warning when not setting a `defaultLocale` when using one of the `_default` strategies in https://github.com/nuxt-modules/i18n/commit/cbb6f22e638f69a71de4eefb51342e7c2e9a06b1, which hopefully prevents further misconfigurations.
